source/defender_demo.c

```diff
@@ -0,0 +1,918 @@
+                 uxStackSize,         /* Size of stack (in words, not bytes) to allocate for the task. */
+                 NULL,                /* Task parameter - not used in this case. */
+                 uxPriority,          /* Task priority, must be between 0 and configMAX_PRIORITIES - 1. */
+                 NULL );              /* Used to pass out a handle to the created task - not used in this case. */
+}
+
+/*-----------------------------------------------------------*/
+
+static bool prvSubscribeToDefenderTopics( void )
+{
+    MQTTStatus_t xStatus;
+    uint32_t ulNotificationValue;
+    CommandInfo_t xCommandParams = { 0 };
+
+    /* These must persist until the command is processed. */
+    static MQTTAgentSubscribeArgs_t xSubscribeArgs;
+    static MQTTSubscribeInfo_t xSubscribeInfo[ 2 ];
+
+    /* Context must persist as long as subscription persists. */
+    static CommandContext_t xApplicationDefinedContext = { 0 };
+
+    /* Record the handle of this task in the context that will be used within
+    * the callbacks so the callbacks can send a notification to this task. */
+    xApplicationDefinedContext.xTaskToNotify = xTaskGetCurrentTaskHandle();
+
+    /* Ensure the return status is not accidentally MQTTSuccess already. */
+    xApplicationDefinedContext.xReturnStatus = MQTTBadParameter;
+
+    /* Subscribe to defender topic for responses for accepted reports. */
+    xSubscribeInfo[ 0 ].pTopicFilter = DEFENDER_API_JSON_ACCEPTED( democonfigCLIENT_IDENTIFIER );
+    xSubscribeInfo[ 0 ].topicFilterLength = DEFENDER_API_LENGTH_JSON_ACCEPTED( democonfigCLIENT_IDENTIFIER_LENGTH );
+    xSubscribeInfo[ 0 ].qos = MQTTQoS1;
+    /* Subscribe to defender topic for responses for rejected reports. */
+    xSubscribeInfo[ 1 ].pTopicFilter = DEFENDER_API_JSON_REJECTED( democonfigCLIENT_IDENTIFIER );
+    xSubscribeInfo[ 1 ].topicFilterLength = DEFENDER_API_LENGTH_JSON_REJECTED( democonfigCLIENT_IDENTIFIER_LENGTH );
+    xSubscribeInfo[ 1 ].qos = MQTTQoS1;
+
+    /* Complete the subscribe information.  The topic string must persist for
+     * duration of subscription - although in this case is it a static const so
+     * will persist for the lifetime of the application. */
+    xSubscribeArgs.pSubscribeInfo = xSubscribeInfo;
+    xSubscribeArgs.numSubscriptions = 2;
+
+    /* Loop in case the queue used to communicate with the MQTT agent is full and
+     * attempts to post to it time out.  The queue will not become full if the
+     * priority of the MQTT agent task is higher than the priority of the task
+     * calling this function. */
+    xTaskNotifyStateClear( NULL );
+    xCommandParams.blockTimeMs = defenderexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = prvSubscribeCommandCallback;
+    xCommandParams.pCmdCompleteCallbackContext = &xApplicationDefinedContext;
+    LogInfo( ( "Sending subscribe request to agent for defender topics." ) );
+
+    do
```

Comment:
Retry infinitely? Isn't return on one failure enough?

Reply:
This is not for failure to subscribe. This fails when the agent's queue is full and we need to try until the agent has processed more items.